### PR TITLE
Add the pyocd as a runner for stm32F303 and stm32f334 nucleo boards

### LINKED ISSUE #54117: West flash fails on upload to Nucleo F303K8
## Bug Description
Flashing of the Nucleo F303K8 on Windows works but under Linux(Ubuntu) fails. I also tested the Nucleo F466RE which works on both systems.

```(.venv) moritz-geier@Galactus:~/Documents/zephyrproject/zephyr$ west flash
-- west flash: rebuilding
ninja: no work to do.
-- west flash: using runner openocd
-- runners.openocd: Flashing file: /home/moritz-geier/Documents/zephyrproject/zephyr/build/zephyr/zephyr.hex
Open On-Chip Debugger 0.11.0+dev-00725-gc5c47943d (2022-11-22-06:32)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : DEPRECATED target event trace-config; use TPIU events {pre,post}-{enable,disable}
Info : clock speed 1000 kHz
Info : STLINK V2J40M27 (API v2) VID:PID 0483:374B
Info : Target voltage: 3.278628
Warn : target stm32f3x.cpu examination failed
Error: jtag status contains invalid mode value - communication failure
Polling target stm32f3x.cpu failed, trying to reexamine
Examination failed, GDB will be halted. Polling again in 100ms
Info : Previous state query failed, trying to reconnect
Error: jtag status contains invalid mode value - communication failure
Polling target stm32f3x.cpu failed, trying to reexamine
Examination failed, GDB will be halted. Polling again in 300ms
Info : starting gdb server for stm32f3x.cpu on 3333
Info : Listening on port 3333 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32f3x.cpu       hla_target little stm32f3x.cpu       unknown

Info : Unable to match requested speed 1000 kHz, using 950 kHz
Info : Unable to match requested speed 1000 kHz, using 950 kHz
Error: read_memory: read at 0xe0042004 with width=32 and count=1 failed
Error executing event examine-end on target stm32f3x.cpu:
/home/moritz-geier/zephyr-sdk-0.15.2/sysroots/x86_64-pokysdk-linux/usr/share/openocd/scripts/mem_helper.tcl:5: Error: read_memory: failed to read memory
in procedure 'ocd_process_reset' 
in procedure 'ocd_process_reset_inner' called at file "embedded:startup.tcl", line 876
in procedure 'stm32f3x_default_examine_end' called at file "/home/moritz-geier/zephyr-sdk-0.15.2/sysroots/x86_64-pokysdk-linux/usr/share/openocd/scripts/target/stm32f3x.cfg", line 100
in procedure 'mmw' called at file "/home/moritz-geier/zephyr-sdk-0.15.2/sysroots/x86_64-pokysdk-linux/usr/share/openocd/scripts/target/stm32f3x.cfg", line 81
in procedure 'mrw' called at file "/home/moritz-geier/zephyr-sdk-0.15.2/sysroots/x86_64-pokysdk-linux/usr/share/openocd/scripts/mem_helper.tcl", line 30
at file "/home/moritz-geier/zephyr-sdk-0.15.2/sysroots/x86_64-pokysdk-linux/usr/share/openocd/scripts/mem_helper.tcl", line 5
Info : Previous state query failed, trying to reconnect
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x08001058 msp: 0x20000fa8
Info : Unable to match requested speed 8000 kHz, using 4000 kHz
Info : Unable to match requested speed 8000 kHz, using 4000 kHz
Polling target stm32f3x.cpu failed, trying to reexamine
Info : [stm32f3x.cpu] Cortex-M4 r0p1 processor detected
Info : [stm32f3x.cpu] target has 6 breakpoints, 4 watchpoints
Info : device id = 0x10016438
Info : flash size = 64kbytes
Warn : Adding extra erase range, 0x080036dc .. 0x080037ff
Error: error writing to flash at address 0x08000000 at offset 0x00000000
auto erase enabled
FATAL ERROR: command exited with status 1: /home/moritz-geier/zephyr-sdk-0.15.2/sysroots/x86_64-pokysdk-linux/usr/bin/openocd -s /home/moritz-geier/Documents/zephyrproject/zephyr/boards/arm/nucleo_f303k8/support -s /home/moritz-geier/zephyr-sdk-0.15.2/sysroots/x86_64-pokysdk-linux/usr/share/openocd/scripts -f /home/moritz-geier/Documents/zephyrproject/zephyr/boards/arm/nucleo_f303k8/support/openocd.cfg '-c init' '-c targets' -c 'reset init' -c 'flash write_image erase /home/moritz-geier/Documents/zephyrproject/zephyr/build/zephyr/zephyr.hex' -c 'reset init' -c 'verify_image /home/moritz-geier/Documents/zephyrproject/zephyr/build/zephyr/zephyr.hex' -c 'reset run' -c shutdown
```

## To Reproduce

west build -p auto -b nucleo_f303k8 samples/basic/blinky
west flash
Expected behavior
To be able to flash the Nucleo F303K8 on Ubuntu with west flash

## Impact
annoyance

## Environment (please complete the following information):
Windows:
- OS: Windows Build 19044
- Toolchain Zephyr SDK 15.2
- openocd 0.11.0 via choco

Ubuntu:
- OS: Ubuntu 22.04
- Toolchain Zephyr SDK 15.2

Firmware on ST-Links are up-to date

### PR DESCRIPTION
**Pull request by FRASTM**
Add the pyocd as a runner for the nucleo_f303k8 and nucleo_f334r8 boards for flashing and debugging.

If the stm32f303k8tx/ stm32f334r8tx  is not in the list of targets for pyocd :
`$ pyocd list --targets`
then add the pack for pyocd
`$ pyocd pack install stm32f303`  `$ pyocd pack install stm32f334`  

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/54117

Nucleo_f334r8 and nucleo_f303k8 report the same error with openocd : unable to write the flash after erasing

Signed-off-by: Francois Ramu <francois.ramu@st.com>

### CODE REVIEW

**Comment by erwango** [comment]
@moritz-geier can you have a try?

**Comment by moritz-geier** [comment]
Since the target stm32f303 is missing I tried to download it, however the packet command `pyocd pack install stm32f303` fails with `0000391 W No matching devices. Please make sure the pack index is up to date. [pack_cmd]` and if i try to run `pyocd pack update` I get
```
0000355 I Updating pack index... [pack_cmd]
08:34:23 [INFO] Downloaded http://www.keil.com/pack/index.pidx
08:34:23 [INFO] Found 833 Pdsc entries
08:34:24 [WARN] No Core found!33)
08:34:24 [WARN] No Core found!
... 44 Times the same message ...
08:34:24 [WARN] No Core found!
08:34:24 [WARN] XML error: Expecting </link> found </head>
08:34:24 [WARN] the end of the document has been reached prematurely
08:34:24 [WARN] Device found without a processor ATSAMDA1E14A
08:34:24 [WARN] Device found without a processor ATSAME70J19
08:34:24 [WARN] Device found without a processor ATSAMS70J19
08:34:24 [WARN] Device found without a processor ATSAMV70J19
08:34:24 [WARN] Device found without a processor ATSAMV71J19
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] cannot parse integer from empty string
08:34:24 [WARN] cannot parse integer from empty string
08:34:24 [WARN] cannot parse integer from empty string
08:34:24 [WARN] cannot parse integer from empty string
08:34:24 [WARN] cannot parse integer from empty string
08:34:24 [WARN] cannot parse integer from empty string
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] Support for two parent processors not implemented!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] Support for two parent processors not implemented!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] No Core found!
08:34:24 [WARN] XML error: Expecting </P> found </BODY>
... 62 Times the same message ...
08:34:24 [WARN] XML error: Expecting </P> found </BODY>
08:34:25 [WARN] Support for two parent processors not implemented!
```
is this related?

**Comment by FRASTM** [comment]
Only if you do not find the stm32f303xx  in the list of your pyocd targets 
(pyocd list --targets or pyocd list --targets | grep stm32f303)  
you should install that pack 
$pyocd pack install stm32f303
should download the pack 
```
Downloading packs (press Control-C to cancel):
    Keil.STM32F3xx_DFP.2.2.2
Downloading descriptors (001/001)
```

**Comment by moritz-geier** [comment]
If I call the `pyocd list --targets | grep stm32f303` nothing returns, thus the target is not installed. When I try to install the package as you said I get the error messages described above. I wrote an [Issue on the pyocd page](https://github.com/pyocd/pyOCD/issues/1496) as soon as this is resolved I will test the changes.

**Comment by FRASTM** [comment]
Adding the stm32CubeProgrammer as a runner for west flash

**Comment by FRASTM** [comment]
> If I call the `pyocd list --targets | grep stm32f303` nothing returns, thus the target is not installed. When I try to install the package as you said I get the error messages described above. I wrote an [Issue on the pyocd page](https://github.com/pyocd/pyOCD/issues/1496) as soon as this is resolved I will test the changes.

For flashing you can still use the stm32cubeprogrammer, but for debugging the pyocd target pack is required.

**Comment by moritz-geier** [comment]
Got it running, I'm now able to flash and debug via pyocd.

**Review by erwango — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

File: soc/arm/st_stm32/stm32f3/soc.c [hunk 1]
```diff
@@ -43,6 +44,9 @@ static int stm32f3_init(const struct device *arg)
 	/* At reset, system core clock is set to 8 MHz from HSI */
 	SystemCoreClock = 8000000;
 
+	/* Allow reflashing the board */
+	LL_DBGMCU_EnableDBGSleepMode();
```
Comment (tagunil) [inline soc/arm/st_stm32/stm32f3/soc.c:48]: What about other STM32 series, wouldn't this change be useful as well?

**Pull request merged**